--- old-glibc-2.3.5/Rules	2004-11-12 01:19:22.000000000 +0000
+++ new-glibc-2.3.5/Rules	2006-07-28 22:37:32.000000000 +0100
@@ -147,7 +147,24 @@
 $(objpfx)%.out: /dev/null $(objpfx)%	# Make it 2nd arg for canned sequence.
 	$(make-test-out) > $@
 
+# Write out arguments associated with test cases to files,
+# so that tests can be used independently of this build system.
+$(objpfx)%.env:
+	echo $($*-ENV) > $@
+$(objpfx)%.args:
+	echo $($*-ARGS) > $@
+$(objpfx)%.cflags:
+	echo $(CFLAGS-$*.c) > $@
+
 endif	# tests
+
+test-list-params: \
+	$(foreach test,$(tests),$(objpfx)$(test).env) \
+	$(foreach test,$(tests),$(objpfx)$(test).args) \
+	$(foreach test,$(tests),$(objpfx)$(test).cflags)
+
+test-list:
+	echo $(tests) > $(objpfx)test-list
 
 .PHONY: distclean realclean subdir_distclean subdir_realclean \
 	subdir_clean subdir_mostlyclean subdir_testclean
